src/chain/shared_state.rs

```diff
@@ -156,6 +157,35 @@ impl SharedState {
                 .collect()
         }
     }
+
+    /// Updates the entry in `their_keys` for `prefix` to the latest known key; if a split
+    /// occurred in the meantime, the keys for sections covering the rest of the address space are
+    /// initialised to the old key that was stored for their common ancestor
+    pub fn update_their_keys(&mut self, prefix: Prefix<XorName>, key: BlsPublicKey) {
+        if let Some(&pfx) = self
+            .their_keys
+            .keys()
+            .find(|pfx| pfx.is_compatible(&prefix))
+        {
+            let old_key = unwrap!(self.their_keys.remove(&pfx));
+            let mut current_pfx = prefix;
+            loop {
+                current_pfx = current_pfx.sibling();
+                if self.their_keys.contains_key(&current_pfx) || self.our_prefix() == &current_pfx {
```

Comment:
I am not 100% sure about that check for our_prefix, is this the symptom of something else?
Should we stop when we reach the prefix we are replacing. (i.e check if it is the old_key_sibling)?

I am also not sure it is correct.
Could you add the test case:

```
    fn simple_split() {
        update_keys_and_check(
            "01", // Not the sibling of the single bit parent prefix of 111
            vec!["1", "111"],
            // ??? 
        );
    }
```

Reply:
Ah, good point. I initially wanted to make that `.is_compatible()` instead of `==`, but I couldn't think of a situation where that would matter - but of course, this is one of them. We should probably do the `is_compatible` thing, then.

Reply:
> Should we stop when we reach the prefix we are replacing. (i.e check if it is the old_key_sibling)?

Reply:
> Should we stop when we reach the prefix we are replacing. (i.e check if it is the old_key_sibling)?

I think this could be an alternative where you can still use ==

Reply:
We would need to check that between `.popped()` and `.sibling()`, though, so the code with the while loop would be slightly uglier (if we could even still use the while loop... I'm not sure). Also, I'm not sure if that's always correct - it probably is, though.

Reply:
No, no need for that, if the thing you are checking for equality is the sibling of the one that was in the map.